### PR TITLE
[CDAP-16290] Instrumented FLL for streaming and batch sources

### PR DESCRIPTION
**Pull request by vravish**
Depends on https://github.com/cdapio/hydrator-plugins/pull/1071

### CODE REVIEW

**Comment by yaojiefeng** [comment]
Let's put the recordLineage in hydrator instead of cdap

**Comment by vravish** [comment]
> Let's put the recordLineage in hydrator instead of cdap

Ack. https://github.com/cdapio/hydrator-plugins/pull/1071

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java [hunk 1]
```diff
@@ -112,4 +126,14 @@ public Schema outputSchema(SalesforceStreamingSourceConfig config) throws Except
     return SalesforceSchemaUtil.getSchema(authenticatorCredentials,
                                           SObjectDescriptor.fromQuery(query));
   }
+
+  private void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema, String operationName,
+                             String description) {
+    LineageRecorder lineageRecorder = new LineageRecorder(context, outputName);
+    lineageRecorder.createExternalDataset(tableSchema);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java]: In streaming source, we don't need to call this, you will need to call https://github.com/cdapio/hydrator-plugins/blob/develop/spark-plugins/src/main/java/io/cdap/plugin/spark/FileStreamingSource.java#L74

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java [hunk 2]
```diff
@@ -112,4 +129,18 @@ public Schema outputSchema(SalesforceStreamingSourceConfig config) throws Except
     return SalesforceSchemaUtil.getSchema(authenticatorCredentials,
                                           SObjectDescriptor.fromQuery(query));
   }
+
+  private void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema,
+                             String operationName, String description)
+    throws DatasetManagementException, TransactionFailureException {
+    Preconditions.checkNotNull(tableSchema, "schema for output %s is null.", outputName);
+    Preconditions.checkNotNull(tableSchema.getFields(), "schema.getFields() for output %s is null.", outputName);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java]: Same here, don't use Precondition, if will fail the pipeline, use if and log a message if schema is null

File: src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java [hunk 3]
```diff
@@ -112,4 +132,24 @@ public Schema outputSchema(SalesforceStreamingSourceConfig config) throws Except
...
+
+  private void recordLineage(StreamingSourceContext context, String outputName, Schema tableSchema,
+                             String operationName, String description)
+    throws DatasetManagementException, TransactionFailureException {
+    if (tableSchema == null) {
+      LOG.warn("Schema for output %s is null. Field-level lineage will not be recorded", outputName);
+      return;
+    }
+    if (tableSchema.getFields() == null) {
+      LOG.warn("schema.getFields() for output %s is null. Field-level lineage will not be recorded", outputName);
```
Comment (yaojiefeng) [inline src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforceStreamingSource.java]: nit - just say schema fields is empty

**Review by yaojiefeng — APPROVED** [review]
minor comment, LGTM

**Pull request merged**